ddtrace/internal/processor/stats.py

```diff
@@ -0,0 +1,239 @@
+from collections import defaultdict
+import threading
+import typing
+
+import ddtrace
+from ddtrace import config
+from ddtrace.vendor.ddsketch.ddsketch import LogCollapsingLowestDenseDDSketch
+from ddtrace.vendor.ddsketch.pb.proto import DDSketchProto
+
+from . import SpanProcessor
+from ...constants import SPAN_MEASURED_KEY
+from .._encoding import packb
+from ..agent import get_connection
+from ..compat import get_connection_response
+from ..compat import time_ns
+from ..hostname import get_hostname
+from ..logger import get_logger
+from ..periodic import PeriodicService
+from ..writer import _human_size
+
+
+if typing.TYPE_CHECKING:
+    from typing import DefaultDict
+    from typing import List
+    from typing import Optional
+
+    from ddtrace import Span
+
+    from ..agent import ConnectionType
+
+
+log = get_logger(__name__)
+
+
+def _is_top_level(span):
+    # type: (Span) -> bool
+    """Return whether the span is a "top level" span."""
+    return (span._local_root is span) or (
+        span._parent is not None and span._parent.service != span.service and span.service is not None
+    )
+
+
+def _is_measured(span):
+    # type: (Span) -> bool
+    """Return whether the span is flagged to be measured or not."""
+    return SPAN_MEASURED_KEY in span.metrics
+
+
+"""
+To aggregate metrics for spans they need to be "uniquely" identified (as
+best as possible). This enables the compression of stat points.
+
+Aggregation can be done using primary and secondary attributes from the span
+stored in a tuple which is hashable in Python.
+"""
+SpanAggrKey = typing.Tuple[
+    str,  # name
+    str,  # service
+    str,  # resource
+    str,  # type
+    int,  # http status code
+    bool,  # synthetics request
+]
+
+
+class SpanAggrStats(object):
+    """Aggregated span statistics."""
+
+    def __init__(self):
+        self.hits = 0
+        self.top_level_hits = 0
+        self.errors = 0
+        self.duration = 0
+        self.ok_distribution = LogCollapsingLowestDenseDDSketch(0.00775, bin_limit=2048)
+        self.err_distribution = LogCollapsingLowestDenseDDSketch(0.00775, bin_limit=2048)
+
+
+def _span_aggr_key(span):
+    # type: (Span) -> SpanAggrKey
+    """Return a hashable key that can be used to uniquely refer to the "same"
+    span.
+    """
+    service = span.service or "unnamed_service"
+    if len(service) > 100:
+        service = service[:100]
+    resource = span.resource or span.name
+    _type = span.span_type
+    if _type and len(_type) > 100:
+        _type = _type[:100]
+
+    status_code = span.metrics.get("http.status_code", None)
+    synthetics = span.context.dd_origin == "synthetics"  # TODO: verify this is the right variant
+    return span.name, service, resource, _type, status_code, synthetics
+
+
+class SpanStatsProcessor(PeriodicService, SpanProcessor):
```

Comment:
```suggestion
class V06SpanStatsProcessor(PeriodicService, SpanProcessor):
```

be sure to match the encoder naming pattern